src/agent.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The Jenkins agent relation observer."""
+import logging
+import typing
+
+from ops.charm import CharmBase, RelationJoinedEvent
+from ops.framework import Object
+from ops.model import ActiveStatus, BlockedStatus, Container, MaintenanceStatus
+
+import jenkins
+from state import State
+
+logger = logging.getLogger(__name__)
+
+
+class AgentRelationData(typing.TypedDict):
+    """Relation data required for adding the Jenkins agent.
+
+    Attrs:
+        url: The Jenkins server url.
+        secret: The secret for agent node.
+    """
+
+    url: str
+    secret: str
+
+
+class Observer(Object):
+    """The Jenkins agent relation observer."""
+
+    def __init__(self, charm: CharmBase, state: State):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+            state: The charm state.
+        """
+        super().__init__(charm, "agent-observer")
+        self.charm = charm
+        self.state = state
+
+        charm.framework.observe(charm.on["agent"].relation_joined, self._on_agent_relation_joined)
+
+    @property
+    def _jenkins_container(self) -> Container:
+        """The Jenkins workload container."""
+        return self.charm.unit.get_container(self.state.jenkins_service_name)
+
+    def _on_agent_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Handle agent relation joined event.
+
+        Args:
+            event: The event fired from an agent joining the relationship.
+        """
+        if not (binding := self.model.get_binding("juju-info")):
+            return
+        if not self._jenkins_container.can_connect():
+            event.defer()
+            return
+        if not event.unit or not all(
+            event.relation.data[event.unit].get(required_data)
+            for required_data in ("executors", "labels", "slavehost")
+        ):
+            logger.warning("Relation data not ready yet. Deferring.")
+            event.defer()
+            return
+        agent_meta = jenkins.AgentMeta(
+            executors=event.relation.data[event.unit]["executors"],
+            labels=event.relation.data[event.unit]["labels"],
+            slavehost=event.relation.data[event.unit]["slavehost"],
+        )
+        try:
+            agent_meta.validate()
+        except jenkins.ValidationError as exc:
+            logger.error("Invalid agent relation data. %s", exc)
+            self.charm.unit.status = BlockedStatus("Invalid agent relation data.")
+            return
+
+        self.charm.unit.status = MaintenanceStatus("Adding agent node.")
+        credentials = jenkins.get_admin_credentials(self._jenkins_container)
+        jenkins_client = jenkins.get_client(client_credentials=credentials)
+        try:
+            jenkins.add_agent_node(
+                jenkins_client=jenkins_client,
```

Comment:
Can Jenkins figure out how to get its own client? I.e., just get `add_agent_node` to call `_get_client`?

Reply:
We would need to pass in the credentials, most of the jenkins functions would either need the credentials or the client. Passing in client seemed more explicit and easier to test.

Reply:
Couldn't those also just be requested by the client creation function?

Carrying around a Jenkins client and passing that around increases coupling between modules so it would be good to minimise if possible.

For testing, the jenkins client can be an optional argument

Reply:
Updated the client to be an optional argument! Thank you :)